platform/duration.cpp

```diff
@@ -0,0 +1,104 @@
+#include "duration.hpp"
+
+namespace platform
+{
+namespace
+{
+using std::chrono::duration_cast, std::chrono::seconds, std::chrono::minutes, std::chrono::hours, std::chrono::days;
+
+long SecondsToUnits(seconds duration, Duration::Units unit)
+{
+  switch (unit)
+  {
+    case Duration::Units::Days: return duration_cast<days>(duration).count();
+    case Duration::Units::Hours: return duration_cast<hours>(duration).count();
+    case Duration::Units::Minutes: return duration_cast<minutes>(duration).count();
+    default: UNREACHABLE();
+  }
+}
+
+seconds UnitsToSeconds(long value, Duration::Units unit)
+{
+  switch (unit)
+  {
+    case Duration::Units::Days: return days(value);
+    case Duration::Units::Hours: return hours(value);
+    case Duration::Units::Minutes: return minutes(value);
+    default: UNREACHABLE();
+  }
+}
+
+std::string GetUnitSeparator(Locale const &locale)
+{
+  static const std::unordered_set<std::string> kEmptyNumberUnitSeparatorLocales =
+  {
+    "en", "de", "fr", "he", "fa", "ja", "ko", "mr", "th", "tr", "vi", "zh"
+  };
+  return kEmptyNumberUnitSeparatorLocales.count(locale.m_language) ? "" : kNarrowNonBreakingSpace;
+}
+
+std::string GetUnitsGroupingSeparator(Locale const & locale)
+{
+  static const std::unordered_set<std::string> kEmptyGroupingSeparatorLocales =
+  {
+    "ja", "zh"
+  };
+  return kEmptyGroupingSeparatorLocales.count(locale.m_language) ? "" : kNonBreakingSpace;
+}
+}
+
+Duration::Duration(long seconds) : m_seconds(seconds)
+{
+  ASSERT_GREATER_OR_EQUAL(seconds, 0, ());
+}
+
+std::string Duration::GetLocalizedString(std::set<Units> units, Locale const & locale) const
+{
+  return GetString(units, GetUnitSeparator(locale), GetUnitsGroupingSeparator(locale));
+}
+
+std::string Duration::GetPlatformLocalizedString() const
+{
+  static auto const kCurrentUnitSeparator = GetUnitSeparator(GetCurrentLocale());
+  static auto const kCurrentGroupingSeparator = GetUnitsGroupingSeparator(GetCurrentLocale());
+  return GetString({Units::Days, Units::Hours, Units::Minutes}, kCurrentUnitSeparator, kCurrentGroupingSeparator);
+}
+
+std::string Duration::GetString(std::set<Units> units, std::string unitSeparator, std::string groupingSeparator) const
+{
+  ASSERT(units.size(), ());
+  if (m_seconds.count() == 0)
+    return std::to_string(0) + unitSeparator + GetUnitsString(Units::Minutes);
+
+  std::string formattedTime;
+  seconds remainingSeconds = m_seconds;
+  
+  for (auto const unit : units)
+  {
+    const long unitsCount = SecondsToUnits(remainingSeconds, unit);
+    if (unitsCount > 0)
+    {
+      if (!formattedTime.empty())
+        formattedTime += groupingSeparator;
+      formattedTime += std::to_string(unitsCount) + unitSeparator + GetUnitsString(unit);
+      remainingSeconds -= UnitsToSeconds(unitsCount, unit);
+    }
+  }
+  return formattedTime;
+}
+
+std::string Duration::GetUnitsString(Units unit)
+{
+  static const std::string kStringsMinute = "minute";
```

Comment:
Prefer using constexpr std::string_view for any string constants. They will be initialized at compile time, not at runtime.